### ISSUE TITLE
Getting sqlalchemy.exc.IntegrityError when executing simultaneous inserts.

### ISSUE THREAD

**Issue by nservidio** [sqlalchemy/sqlalchemy#5083]
Hi,

Thank you so much for creating SQLAlchemy!

### Summary

We are currently seeing the following error:
```
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(mysql.connector.errors.IntegrityError) 1062 (23000): Duplicate entry '1-62' for key 'user_id_item_id'
[SQL: INSERT INTO item_user_associations (user_id, item_id, extra_data) VALUES (%(user_id)s, %(item_id)s, %(extra_data)s)]
[parameters: {'user_id': 1, 'item_id': 62, 'extra_data': 'hello world!'}]
(Background on this error at: http://sqlalche.me/e/gkpj)
```
We believe the error occurs when users send two POST requests to our API containing similar data in very close succession. However, the issue is reproducible without networking.

Here is an example PyCharm project to reproduce the issue:
[AssociationErrorExample-test.zip](https://github.com/sqlalchemy/sqlalchemy/files/4032490/AssociationErrorExample-test.zip)

### Steps to Repro

1. Get a local copy of MySQL running:
    ```
     docker run --name my_db \
          -e MYSQL_DATABASE=my_db \
          -e MYSQL_USER=user \
          -e MYSQL_PASSWORD=password \
          -p 3306:3306 -d mysql/mysql-server:5.7
    ```
1. Run the "Compound" run configuration from PyCharm:
    ```
    Run > Run 'Compound'
    ```
   This executes the "example" run configuration 8 times in parallel.   
1. As an alternative you can execute example.py in multiple terminal windows.

Expected: No error.

Actual: You see an `sqlalchemy.exc.IntegrityError` error in at least one of the outputs.

### Workaround

We've been doing something like this to overcome this issue:

```
@contextmanager
def session_scope():
    """
    This code is taken from the final example here:
    https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it
    """
    session = Session()
    try:
        yield session
        session.commit()
    except exc.InternalError:
        session.rollback()
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()
```

### Questions

Is there a built in SQLAlchemy mechanism or recommended best practice to overcome this issue?

Please let me know. Thanks!

-Nick

**Comment by zzzeek** [comment]
hi there -

there's not much mysterious about this error message as it is the required outcome of an application that tries to INSERT duplicate records from multiple processes.    how to handle it depends on what you actually need to do.   There are many possiblitiies, some are below and I've bulleted them out:

* if the error is because the user presses submit twice, then the second request should be ignored, in which case you'd catch the integrity error, possibly log it, and then discard the request.    

* if the error is because two different users are trying to insert the same record at the same time, you might want to consider:
   * a more concurrent-capable datamodel, like one where the actions by a particular user are also keyed to that user, or to a timestamp, 
  * or again where the possible contention for the row is anticipated, you catch the integrity error and then notify the user that someone else inserted that record already.

* still another way to go is to use an "upsert", for MySQL we support it via https://docs.sqlalchemy.org/en/13/dialects/mysql.html#insert-on-duplicate-key-update-upsert ; this is Core so you'd have to run this particular operation without using the ORM for the persistence part.    

* another way to do upsert that is more ORM friendly is to run the insert inside of a savepoint, then if integrity error, end the savepoint and then do an UPDATE instead: https://docs.sqlalchemy.org/en/13/orm/session_transaction.html?highlight=savepoint#using-savepoint  

* all the above methods are known as "optimistic", where they expect success and adapt if failure happens.  There is also "pessimistic", where you anticipate a failure typically using locking.  In this case since INSERT is the issue you could possibly use SERIALIZABLE isolation for this particular operation, such that as long as your operation does a SELECT for the row first, it will never see the row that does not exist suddenly appear when it tries to do an INSERT.  This would reduce the concurrency of your application within the scope of this operation, however if this is a particular operation that can afford to be this way, like creating a username, it may be worth it.  Turning on per-transaction SERIALIZABLE is documented at https://docs.sqlalchemy.org/en/13/orm/session_transaction.html?highlight=savepoint#setting-isolation-for-individual-sessions


now for all of that advice, what I can't do here is run pycharm or open zip files or anything like that, the issue tracker works on the Stack Overflow model of people posting [MCVEs](http://stackoverflow.com/help/mcve) to illustrate specific API problems.  This is a more general application design issue and you may have more luck getting folks to open up your application to review on the mailing list or IRC channel, see https://www.sqlalchemy.org/support.html for background on that.

**Comment by zzzeek** [comment]
reopen if you have further questions thanks!

**Comment by advance512** [comment]
@zzzeek 
Are there plans to support "upsert" operations for MySQL (and other databases that support it, e.g. sqlite) in the ORM, as opposed to Core?

I have spent hours and found no simple way to add support to a single ORM model to do "upserts" instead of normal inserts. Everything requires dropping down to Core, or performing global changes to all models.

**Comment by zzzeek** [comment]
> @zzzeek
> Are there plans to support "upsert" operations for MySQL (and other databases that support it, e.g. sqlite) in the ORM, as opposed to Core?
> 
> I have spent hours and found no simple way to add support to a single ORM model to do "upserts" instead of normal inserts. Everything requires dropping down to Core, or performing global changes to all models.

people have asked for this for a long time but it's not on the roadmap right now.    my rough understanding of the various "upsert" options is that they behave pretty differently on each backend, both in what kinds of options can be set up for row conflicts as well as how SQLAlchemy's "defaults" would behave, we have three kinds of defaults for both inserts and updates that people would report as bugs if they didn't behave the same in an "upsert" scenario, none of that is implemented right now.

I think the last time I thought about this, it wouldn't be merged into the base "flush()" process it will still be in some kind of session.merge() method, but obviously it would be most useful if there were some kind of "mass_merge()" thing.  the issue of how to integrate Python-side "oninsert"and "onupdate" defaults was one of my main concerns.

When I need to merge things and I have the data in memory already anyway, I usually just use a chunked SELECT -> INSERT/UPDATE approach, just select all the IDs that we have already and work from there, works consistently on all backends and performs pretty similarly too.

**Comment by zzzeek** [comment]
more directly I haven't yet seen why SQL-side "upsert" on a plain primary key and nothing else, where I have the data on the python side already, is that important (noting that this is how I am with *everything* at first, always takes me awhile to see why something is useful).    I should add the recipes I use for this to the wiki if I haven't already.

**Comment by advance512** [comment]
@zzzeek 

Well, as for your why SQL-side "upserts" are important:

1. I guess I don't need to explain why "upserts" are helpful in general, as they are supported in SQLAlchemy Core.
This article discusses _some_ of the reason why "upserts" are useful: https://lucumr.pocoo.org/2014/2/16/a-case-for-upserts/
"Upserts" basically add a SET mode of operation, in addition to INSERT and UPDATE.

2. "Upserts" are supported in SQLAlchemy Core, so having feature parity (or as near it as can get) with ORM makes sense to me, from a product perspective. 

3. Imagine a situation where a website is scraped. You have many workers that receive URLs to scrape, and some of them might receive the same URL to scrape, for various possible reasons.  In such a situation, you want to write the scraped results to the database, be it an INSERT or an UPDATE - it is actually more like a SET operation. You don't care about older results, whether they exist or do not - you just want to write the latest results to the database. To do this now, you'd have to perform a SELECT query to see if there's a result already, to decide whether to do an INSERT or an UPDATE. Then you can INSERT or UPDATE. This causes this operation to take a minimum of two round-trip times, instead of one.  
In the case of an INSERT, if during this time, someone else INSERTs the data - you'll get an error. Now, you can say that having an error here is fine, since maybe the older data is sufficient, it isn't that much older. But the older data might be a "tombstone", maybe the scraping failed and your new data is actually better - but it'll get lost. Or, you'll be forced to write error handling code, and changing the operation to an UPDATE, which puts you in 3 round-trip times instead of 1. And of course, errors occurring here make code more complex, adding to cognitive load and making code less maintainable.
You could also SELECT .. FOR UPDATE, though I am not sure if this would help in the case of an INSERT. Even if it did help, it would require more strict isolation levels for all connections to the database, which surely would hamper performance.
If you use SQLAlchemy ORM here, instead of SQLAlchemy Core (shall we say - because you use ORM everywhere else), then you are fresh out of luck. "Upserts" would be an productive and helpful tool, and one way (among several others) to solve the problem, but they are not supported.

There's a high probability I am missing something here, as I don't work in this domain in my day to day, so your thoughts would be appreciated.

**Comment by zzzeek** [comment]
so my usual backend agnostic approach is to SELECT the ids that are in the database ahead, chunking if necessary, then emit out INSERTs for the ones that aren't there and UPDATE for the ones that are.

the actual scenario where we're racing other processes to INSERT a row is a special case and one where we might want to use the Core upsert, sure.     That's why it's there.    

The reason ORM is not a priority is because the behavioral contract of the ORM is more elaborate than this, specifically in that the Core insert()/update() constructs as well as the way the unit of work flush works around them is that we can run default-generation functions that occur for INSERT statements and UPDATE statements separately.   When the statement is an "upsert", we can't run these, because for a given row, we don't know ahead of time if it will be an insert or an update.   There are likely some other subtle behaviors that differ between INSERT and UPDATE that would have to be rethought in the context of a statement that can be either one.   this is why "upsert" out of the gate for the ORM would have to be within a special method that resembles "merge()", rather than integrated into the standard flush process.

The "upsert" also does not work on any backend except MySQL and PostgreSQL, using entirely different structural semantics.   

I am fully in support of some kind of session.merge_upsert() method that actually does this, but it is not easy to do without making mistakes in the API design and we already have a means of people making use of upserts, so it's just not something I have time to work on right now.

**Comment by zzzeek** [comment]
I actually would want to address the plain session.merge() as well and provide a method that takes a list of objects, and if upsert is not available, it will do a SELECT for all the rows en-mass using IN.    So a new form of merge() should be added.

**Comment by zzzeek** [comment]
added this at #5441
